### PR TITLE
Add workflow to restore Kafka from snapshots

### PR DESCRIPTION
**Pull request by cirego**
This script assumes we only care about `key`, `value`, and `timestamp` for each message in Kafka. For each message in a topic, it preserves the original value of these in an Apache Arrow file. Arrow was chosen as a handy method for serializing / deserializing data that also works with many other tools in the Python ecosystem.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4929)

### CODE REVIEW

**Comment by cirego** [comment]
This is the error message I'm getting when trying to run the `snapshot_bench` workflow:

`Dec 03 09:05:17.611 ERROR dataflow::decode::avro: avro deserialization error: schema not found`

**Comment by cirego** [comment]
The schemas exist in the schema registry:

```
[chris-work:~/materialize/materialize/misc/kafka-util:(snap-restore-topics[|]*)] curl -s http://localhost:8081/subjects/debezium.tpcch.item-key/versions/latest | jq .
{
  "subject": "debezium.tpcch.item-key",
  "version": 1,
  "id": 11,
  "schema": "{\"type\":\"record\",\"name\":\"Key\",\"namespace\":\"debezium.tpcch.item\",\"fields\":[{\"name\":\"i_id\",\"type\":\"int\"}],\"connect.name\":\"debezium.tpcch.item.Key\"}"
}
[chris-work:~/materialize/materialize/misc/kafka-util:(snap-restore-topics[|]*)] curl -s http://localhost:8081/subjects/debezium.tpcch.item-value/versions/latest | jq .
{
  "subject": "debezium.tpcch.item-value",
  "version": 1,
  "id": 12,
  "schema": "{\"type\":\"record\",\"name\":\"Envelope\",\"namespace\":\"debezium.tpcch.item\",\"fields\":[{\"name\":\"before\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Value\",\"fields\":[{\"name\":\"i_id\",\"type\":\"int\"},{\"name\":\"i_im_id\",\"type\":[\"null\",{\"type\":\"int\",\"connect.type\":\"int16\"}],\"default\":null},{\"name\":\"i_name\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"i_price\",\"type\":[\"null\",{\"type\":\"bytes\",\"scale\":2,\"precision\":5,\"connect.version\":1,\"connect.parameters\":{\"scale\":\"2\",\"connect.decimal.precision\":\"5\"},\"connect.name\":\"org.apache.kafka.connect.data.Decimal\",\"logicalType\":\"decimal\"}],\"default\":null},{\"name\":\"i_data\",\"type\":[\"null\",\"string\"],\"default\":null}],\"connect.name\":\"debezium.tpcch.item.Value\"}],\"default\":null},{\"name\":\"after\",\"type\":[\"null\",\"Value\"],\"default\":null},{\"name\":\"source\",\"type\":{\"type\":\"record\",\"name\":\"Source\",\"namespace\":\"io.debezium.connector.postgresql\",\"fields\":[{\"name\":\"version\",\"type\":\"string\"},{\"name\":\"connector\",\"type\":\"string\"},{\"name\":\"name\",\"type\":\"string\"},{\"name\":\"ts_ms\",\"type\":\"long\"},{\"name\":\"snapshot\",\"type\":[{\"type\":\"string\",\"connect.version\":1,\"connect.parameters\":{\"allowed\":\"true,last,false\"},\"connect.default\":\"false\",\"connect.name\":\"io.debezium.data.Enum\"},\"null\"],\"default\":\"false\"},{\"name\":\"db\",\"type\":\"string\"},{\"name\":\"schema\",\"type\":\"string\"},{\"name\":\"table\",\"type\":\"string\"},{\"name\":\"txId\",\"type\":[\"null\",\"long\"],\"default\":null},{\"name\":\"lsn\",\"type\":[\"null\",\"long\"],\"default\":null},{\"name\":\"xmin\",\"type\":[\"null\",\"long\"],\"default\":null}],\"connect.name\":\"io.debezium.connector.postgresql.Source\"}},{\"name\":\"op\",\"type\":\"string\"},{\"name\":\"ts_ms\",\"type\":[\"null\",\"long\"],\"default\":null},{\"name\":\"transaction\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"ConnectDefault\",\"namespace\":\"io.confluent.connect.avro\",\"fields\":[{\"name\":\"id\",\"type\":\"string\"},{\"name\":\"total_order\",\"type\":\"long\"},{\"name\":\"data_collection_order\",\"type\":\"long\"}]}],\"default\":null}],\"connect.name\":\"debezium.tpcch.item.Envelope\"}"
}
```

**Comment by cirego** [comment]
Talked with @umanwizard and @ruchirK: The avro messages in Kafka are prepended with the ID of the schema from schema registry. The snapshot / restore code does not maintain the schema IDs and that is likely the bug here.

**Comment by umanwizard** [comment]
@cirego does this require my review? (Just asking since it's still marked as draft)

**Comment by cirego** [comment]
> @cirego does this require my review? (Just asking since it's still marked as draft)

@umanwizard I am cleaning this up now. Okay if I ping you once it's ready?

**Comment by umanwizard** [comment]
Works for me.

**Comment by cirego** [comment]
@quodlibetor and @umanwizard, this is ready for review if you're willing to review it.

I do have an open issue #4994 to fix a possible data quality issue with measurement. Ideally, I'd like this review to cover the workflow and framework, as we are still evolving the actual benchmark.

File: demo/chbench/bin/generate_snapshot [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+export MZ_THREADS=24
+export NUM_WAREHOUSES=10
+export OLAP_THREADS=1
+export OLTP_THREADS=64
+export MZ_TOPIC_SNAPSHOTS=/home/chris/snapshots3
```
[the remark this answers is not in the document]
Comment (cirego) [reply demo/chbench/bin/generate_snapshot:15]: This was mainly for testing purposes. Now that the workflows are working properly, I'll make this parameterized.

File: misc/kafka-util/archive_topics.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        assert message.partition == 0, f"Expected partition 0, got {message.partion}"
+        assert (
+            message.timestamp_type == 0
+        ), "Expected timestamp to be 0 (CreateTime) but got {message.timestamp_type}"
+        assert (
+            message.headers == []
+        ), f"Expected empty list of headers, got {message.headers}"
+        assert (
+            message.serialized_header_size == -1
+        ), f"Expected negative serialized header size, got {message.serialized_header_size}"
```
[the remark this answers is not in the document]
Comment (cirego) [reply misc/kafka-util/archive_topics.py:93]: Thanks, I'll fix the message.

**Comment by umanwizard** [comment]
@quodlibetor are you planning on reviewing this today, or should I?

**Review by quodlibetor — APPROVED** [review]
LGTM!

File: misc/kafka-util/Dockerfile [hunk 3]
```diff
@@ -7,10 +7,16 @@
 # the Business Source License, use of this software will be governed
 # by the Apache License, Version 2.0.
 
-FROM python:3.9-slim
+FROM python:3.7-slim
```
[the remark this answers is not in the document]
Comment (cirego) [reply misc/kafka-util/Dockerfile:10]: Sure!

File: misc/kafka-util/scripts/archive_topics.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        assert message.partition == 0, f"Expected partition 0, got {message.partion}"
+        assert (
+            message.timestamp_type == 0
+        ), "Expected timestamp to be 0 (CreateTime) but got {message.timestamp_type}"
+        assert (
+            message.headers == []
+        ), f"Expected empty list of headers, got {message.headers}"
+        assert (
+            message.serialized_header_size == -1
+        ), f"Expected negative serialized header size, got {message.serialized_header_size}"
```
[the remark this answers is not in the document]
Comment (cirego) [reply misc/kafka-util/scripts/archive_topics.py:96]: In this case, I consider this to be an combined invariant of the test case and data it is generating. In one way, it's testing a program invariant about it's other half. What would you prefer to see? Raising an error, exiting with an exit code, logging an error, something else?

**Pull request merged**